models/yolo/yolo_impl.hpp

```diff
@@ -0,0 +1,168 @@
+/**
+ * @file yolo_impl.hpp
+ * @author Kartik Dutt
+ *
+ * Implementation of YOLO models using mlpack.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MODELS_YOLO_IMPL_HPP
+#define MODELS_YOLO_IMPL_HPP
+
+#include "yolo.hpp"
+
+namespace mlpack {
+namespace ann {
+
+template<
+    typename OutputLayerType,
+    typename InitializationRuleType,
+    std::string YOLOVersion = "v1-tiny"
+>
+YOLO<OutputLayerType, InitializationRuleType, YOLOVersion>::YOLO() :
+    inputChannel(0),
+    inputWidth(0),
+    inputHeight(0),
+    numClasses(0),
+    numBoxes(0),
+    featureWidth(0),
+    featureHeight(0),
+    weights("none")
+{
+  // Nothing to do here.
+}
+
+template<
+    typename OutputLayerType,
+    typename InitializationRuleType,
+    std::string YOLOVersion = "v1-tiny"
+>
+YOLO<OutputLayerType, InitializationRuleType, YOLOVersion>::YOLO(
+  const size_t inputChannel,
+  const size_t inputWidth,
+  const size_t inputHeight,
+  const size_t numClasses,
+  const size_t numBoxes,
+  const size_t featureWidth,
+  const size_t featureHeight,
+  const std::string& weights,
+  const bool includeTop) :
+  YOLO<OutputLayerType, InitializationRuleType, YOLOVersion>(
+    std::tuple<size_t, size_t, size_t>(
+      inputChannel,
+      inputWidth,
+      inputHeight),
+      numClasses,
+      numBoxes,
+      std::tuple<size_t, size_t>(featureWidth, featureHeight),
+      weights,
+      includeTop)
+{
+  // Nothing to do here.
+}
+
+template<
+     typename OutputLayerType,
+     typename InitializationRuleType,
+     std::string YOLOVersion
+>
+YOLO<OutputLayerType, InitializationRuleType, YOLOVersion>::YOLO(
+    const std::tuple<size_t, size_t, size_t> inputShape,
+    const size_t numClasses,
+    const size_t numBoxes,
+    const std::tuple<size_t, size_t> featureShape,
+    const std::string& weights,
+    const bool includeTop) :
+    inputChannel(std::get<0>(inputShape)),
+    inputWidth(std::get<1>(inputShape)),
+    inputHeight(std::get<2>(inputShape)),
+    numClasses(numClasses),
+    numBoxes(numBoxes),
+    featureWidth(std::get<0>(featureShape)),
+    featureHeight(std::get<1>(featureShape)),
+    weights(weights)
+{
+  std::set<string> supportedVersion({"v1-tiny"});
+  mlpack::Log::Assert(supportedVersion.count(YOLOVersion),
+      "Unsupported YOLO version. Trying to find :", YOLOVersion);
+
+  if (weights == "voc")
+  {
+    // Download weights here.
+    LoadModel("./../weights/YOLO/yolo" + YOLOVersion + "_voc.bin");
```

Comment:
I have added a simple tests in this PR. I haven't added tests for weights because of mlpack/mlpack#2561. I think we can merge this PR since the model is working and any changes would be internal implementation in mlpack. To see if the model is working correctly you can run the following commands :
```
git clone https://github.com/kartikdutt18/mlpack-PyTorch-Weight-Translator.git
cd mlpack-PyTorch-Weight-Translator
./run.sh yolov1_tiny
```
With this you will see the model weights being created and transferred into the YOLO model and being used for prediction. You can see that the iou of predictions made by PyTorch model and YOLO model are same i.e. 1.0. Kindly let me know if you face any issues. If this makes sense, I think we can merge this PR and then later add a PR with tests for weights once mlpack/mlpack#2561 is resolved.
(Similar comment from DarkNet).

Reply:
With this we should also merge #29 and #25 as well.

Reply:
@kartikdutt18 Rather than creating a weight file, how about save it to the `weight` directory for making it ready-for-use?

Reply:
This would face the same problem as DarkNet since this also uses BatchNorm Layer which has a serialization issue.